### ISSUE TITLE
Question: Fastest (most efficient way) to store/retrieve WebAssembly(Wasm) binary format

### ISSUE THREAD

**Issue by tpmccallum** [mysqljs/mysql#2360]
Hi,
I am currently storing/retrieving WebAssembly(Wasm) using this `mysqljs/mysql` application, as part of [a lightweight Node.js application for deploying and executing WebAssembly(Wasm) Binary Format via HTTP](https://github.com/second-state/wasm-joey).

## Storage
The database table uses `LONGBLOB`
The Node.js code converts the [Wasm Binary Format](https://webassembly.github.io/spec/core/binary/index.html) to `Uint8Array` and stores it

## Retrieval
It appears that when the Node.js code retrieves the originally stored `Uint8Array` from that MySQL column it is in the form of a hex string Buffer. 

**I want to see if anyone knows of the fastest most efficient way to get the data back to the original Uint8Array. This approach is starting to look like it needs additional processing which may slow things down a little in terms of speed** 

**Question**
It may be possible to store the Wasm on `fs` and just store the path in the DB. However, having MySQL manage who can access what etc. is attractive (file system adds another dimension of access control/user management etc.)

What would the most efficient retrieval of the original Uint8Array from within the LONGBLOB look like?

Thanks in advance.

**Comment by tpmccallum** [comment]
Just a bit of added information
If the data is just converted from the mysql LONGBLOB to `Uint8Array` it is like this
```
Uint8Array(146450) [
  48, 44, 57, 55, 44, 49, 49, 53, 44, 49, 48, 57,
  44, 49, 44, 48, 44, 48, 44, 48, 44, 49, 44, 56,
  54, 44, 49, 52, 44, 57, 54, 44, 50, 44, 49, 50,
  55, 44, 49, 50, 55, 44, 49, 44, 49, 50, 55, 44,
  57, 54, 44, 51, 44, 49, 50, 55, 44, 49, 50, 55,
  44, 49, 50, 55, 44, 49, 44, 49, 50, 55, 44, 57,
  54, 44, 49, 44, 49, 50, 55, 44, 48, 44, 57, 54,
  44, 50, 44, 49, 50, 55, 44, 49, 50, 55, 44, 48,
  44, 57, 54, 44,
  ... 146350 more items
]
```

Whereas what I actually want is as follows
```
Uint8Array(52267) [
    0,  97, 115, 109,   1,   0,   0,   0,   1,  86,  14,  96,
    2, 127, 127,   1, 127,  96,   3, 127, 127, 127,   1, 127,
   96,   1, 127,   0,  96,   2, 127, 127,   0,  96,   1, 127,
    1, 127,  96,   3, 127, 127, 127,   0,  96,   4, 127, 127,
  127, 127,   0,  96,   1, 127,   1, 126,  96,   4, 127, 127,
  127, 127,   1, 127,  96,   5, 127, 127, 127, 127, 127,   0,
   96,   0,   0,  96,   0,   1, 127,  96,   6, 127, 127, 127,
  127, 127, 127,   1, 127,  96,   3, 126, 127, 127,   1, 127,
    2,  35,   1,  22,
  ... 52167 more items
]
```

Smaller and literally the original data which I put into the DB at the outset.

Thanks again, much appreciated.

**Comment by dougwilson** [comment]
Is it possible to show the code that you are using to do the store and load using this module?

**Comment by tpmccallum** [comment]
Hi @dougwilson 
Thank you for the quick reply, I really appreciate it!

This is the single JS file that I am using [1]

This is an example of where I convert the Wasm binary format to Uint8Array for storage in MySQL [2]

This is an example of where I retrieve the Uint8Array from MySQL [3]

Ultimately I need the data as Uint8Array (because that is what the VM which this application calls requires). I am striving for the lowest overhead in terms of performance and speed. You will see that at present I am going to and from String. This works but I would prefer to pass in Uint8Array (not Uint8Array.toString()) if it were faster. Only issue is I am not sure how to get my Uint8Array back, at the point where I retrieve [3] from mysqljs/mysql 

When using LONGBLOB column, the retrieved data seems to introduce a new data type to my code  i.e.hex string Buffer and I wonder if this Buffer will be more of a performance overhead (to decode) than just using LONGTEXT [4] and storing/loading the Uint8Array as non-binary string (using.toString() and split(“,”) as mentioned above)

Thanks so much, any and all suggestions are welcome.

Kind regards 
Tim

[1] https://github.com/second-state/wasm-joey/blob/master/src/server.js#L47
[2] https://github.com/second-state/wasm-joey/blob/master/src/server.js#L357
[3] https://github.com/second-state/wasm-joey/blob/master/src/server.js#L696
[4] https://github.com/second-state/wasm-joey/blob/master/documentation/installation.md#mysql-setup-for-application

**Comment by tpmccallum** [comment]
Interestingly the following command which creates a Uint8Array directly from the data which the DB returns takes a couple of ms longer to perform
```
let view = new Uint8Array(db_buffer.buffer, db_buffer.byteOffset, db_buffer.byteLength / Uint8Array.BYTES_PER_ELEMENT);
```
14ms

**Comment by tpmccallum** [comment]
It turns out that the conversion from Buffer toString is not actually the bottle neck. The `toString()` conversion only takes about 1ms.

I learned today that the `split(",")` actually takes the time from 1ms up to 12ms.

I will look at another method of performing the split. I already tried a ready-made split library but it doubled the time taken (I guess it was built for ease of use etc. and not performance per se)
